File: backend/internal/model/transaction.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

// Transaction kinds (ADR-0017). `kind` classifies a row for both quantity
// reconstruction (positions = Σ amount per account/asset) and flow analytics.
// Only KindFlow is counted by spending / cash-flow / budget aggregates;
// KindOpeningBalance and KindAdjustment are quantity facts but not flow.
const (
	KindFlow           = "flow"
	KindTradeLeg       = "trade_leg"
	KindOpeningBalance = "opening_balance"
	KindAdjustment     = "adjustment"
)

// Transaction is a movement of `Amount` units of `AssetID` in or out of
// `AccountID`. Per ADR-0013, `Amount` is the quantity (positive = in,
// negative = out) and `AssetID` is the unit. For cash transactions
// `AssetID` equals the parent account's primary_quote_asset. For trades
// (issue #238), the cash leg and security leg each carry their own
// `AssetID` and a shared `TransferPairID`. `Kind` (ADR-0017) classifies
// the row; it defaults to KindFlow.
type Transaction struct {
	ID                   int64           `gorm:"primaryKey" json:"id"`
	UserID               int64           `gorm:"not null" json:"user_id"`
	AccountID            int64           `gorm:"not null" json:"account_id"`
	AssetID              int64           `gorm:"not null" json:"asset_id"`
	CategoryID           *int64          `json:"category_id,omitempty"`
	Kind                 string          `gorm:"not null;default:flow" json:"kind"`
	Amount               decimal.Decimal `gorm:"type:numeric(30,18);not null" json:"amount"`
	Description          *string         `json:"description,omitempty"`
	DescriptionClean     *string         `json:"description_clean,omitempty"`
	MerchantName         *string         `json:"merchant_name,omitempty"`
	TransactionDate      time.Time       `gorm:"type:date;not null" json:"transaction_date"`
	PostedDate           *time.Time      `gorm:"type:date" json:"posted_date,omitempty"`
	Source               string          `gorm:"not null" json:"source"`
	ExternalID           *string         `gorm:"column:external_id" json:"external_id,omitempty"`
	PlaidTransactionID   *string         `gorm:"column:plaid_transaction_id" json:"plaid_transaction_id,omitempty"`
	CategorizationMethod *string         `json:"categorization_method,omitempty"`
	CategorizationRuleID *int64          `json:"categorization_rule_id,omitempty"`
	IsTransfer           bool            `gorm:"not null;default:false" json:"is_transfer"`
	TransferPairID       *int64          `json:"transfer_pair_id,omitempty"`
	Notes                *string         `json:"notes,omitempty"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
	DeletedAt            gorm.DeletedAt  `gorm:"index" json:"-"`

	Account      *Account     `gorm:"foreignKey:AccountID" json:"-"`
	Asset        *Asset       `gorm:"foreignKey:AssetID" json:"-"`
	Category     *Category    `gorm:"foreignKey:CategoryID" json:"-"`
	TransferPair *Transaction `gorm:"foreignKey:TransferPairID" json:"-"`
}

func (Transaction) TableName() string { return "transactions" }

File: frontend/src/types/transaction.ts
// Transaction mirrors backend model.Transaction. Money fields are decimal
// strings; format via AmountDisplay (never parse to Number). Per ADR-0013,
// the transaction's unit is `asset_id` — for cash legs it equals the
// parent account's primary_quote_asset; for trade security legs it's the
// security (AAPL, BTC, …).
export type Transaction = {
  id: number
  user_id: number
  account_id: number
  asset_id: number
  category_id?: number | null
  amount: string
  description?: string | null
  description_clean?: string | null
  merchant_name?: string | null
  transaction_date: string  // YYYY-MM-DD (DATE column)
  posted_date?: string | null
  source: TransactionSource
  external_id?: string | null
  plaid_transaction_id?: string | null
  categorization_method?: string | null
  is_transfer: boolean
  transfer_pair_id?: number | null
  notes?: string | null
  created_at: string
  updated_at: string
}

export const TRANSACTION_SOURCES = ['manual', 'plaid', 'csv', 'pdf'] as const
export type TransactionSource = (typeof TRANSACTION_SOURCES)[number]

// CreateTransactionInput mirrors backend handler.createTransactionRequest.
// Per ADR-0013, no `currency` field — the asset is derived server-side
// from the parent account.
export type CreateTransactionInput = {
  account_id: number
  category_id?: number | null
  amount: string
  description?: string | null
  merchant_name?: string | null
  transaction_date: string  // YYYY-MM-DD accepted
  posted_date?: string | null
  source?: TransactionSource
  notes?: string | null
}

// ── CSV import (#330) — mirrors backend service.ImportResult ──

// ColumnMapping ties the three logical fields to source-CSV header names.
export type ColumnMapping = {
  date: string
  amount: string
  description: string
}

export type ImportRowStatus = 'new' | 'duplicate' | 'error'

export type ImportRowResult = {
  line: number
  date: string
  amount: string
  description: string
  external_id?: string
  status: ImportRowStatus
  error?: string
  // confidence is the extractor's certainty (0–1). Deterministic CSV rows are
  // 1.0; the AI extractor (ADR-0019 phase 2) reports real values. needs_review
  // flags low-confidence new rows the UI surfaces before commit.
  confidence: number
  needs_review: boolean
}

export type ImportResult = {
  committed: boolean
  source: string
  mapping: ColumnMapping
  headers: string[]
  total_rows: number
  new_count: number
  duplicate_count: number
  error_count: number
  review_count: number
  inserted_count: number
  rows: ImportRowResult[]
}

// UpdateTransactionInput: sparse patch. `clear_category: true` with a null
// `category_id` uncategorizes; a non-null `category_id` sets it; absent both
// leaves the category alone.
export type UpdateTransactionInput = {
  category_id?: number | null
  clear_category?: boolean
  amount?: string
  description?: string | null
  merchant_name?: string | null
  transaction_date?: string
  posted_date?: string | null
  notes?: string | null
}
